docker/docker_image_dest.go

```diff
@@ -596,6 +637,125 @@ func (d *dockerImageDestination) putOneSignature(url *url.URL, signature []byte)
 	}
 }
 
+func (d *dockerImageDestination) putSignaturesToCosignAttachments(ctx context.Context, signatures []signature.Cosign, manifestDigest digest.Digest) error {
+	if !d.c.useCosignAttachments {
+		return errors.New("writing Cosign attachments is disabled by configuration")
+	}
+
+	ociManifest, err := d.c.getCosignAttachmentManifest(ctx, d.ref, manifestDigest)
+	if err != nil {
+		return nil
+	}
+	var ociConfig imgspecv1.Image // Most fields empty by default
+	if ociManifest == nil {
+		ociManifest = manifest.OCI1FromComponents(imgspecv1.Descriptor{
+			MediaType: imgspecv1.MediaTypeImageConfig,
+			Digest:    "", // We will fill this in later.
+			Size:      0,
+		}, nil)
+	} else {
+		configBlob, err := d.c.getOCIDescriptorContents(ctx, d.ref, ociManifest.Config, iolimits.MaxConfigBodySize,
+			none.NoCache) // FIXME: Pass BlobInfoCache to signature accessors?)
+		if err != nil {
+			return err
+		}
+		if err := json.Unmarshal(configBlob, &ociConfig); err != nil {
+			return fmt.Errorf("parsing Cosign config %s in %s: %w", ociManifest.Config.Digest.String(),
+				d.ref.ref.Name(), err)
+		}
+	}
+
+	for _, sig := range signatures {
+		found := false
+		for _, layer := range ociManifest.Layers {
+			if cosignSignatureMatchesLayer(sig, layer) {
+				found = true
+				break
+			}
+		}
+		if found {
+			continue
+		}
+
+		sigDesc, err := d.putBlobBytesAsOCI(ctx, sig.UntrustedPayload, sig.UntrustedMIMEType, private.PutBlobOptions{
+			Cache:      none.NoCache, // FIXME: Pass BlobInfoCache to signature accessors? Or is this just pointless noise??
+			IsConfig:   false,
+			EmptyLayer: false,
+			LayerIndex: nil,
+		})
+		if err != nil {
+			return err
+		}
+		sigDesc.Annotations = sig.UntrustedAnnotations
```

Comment:
I _think_ that using `sig` is dangerous.  If there are multiple signatures, wouldn't all users of `UntrustedAnnotations` point to the field of the last item in the slice?

Reply:
Nevertheless, doing more explicit copies when creating/consuming signatures would be worth the peace of mind.